src/enable_variable_ops/ngraph_capture_variables.cc

```diff
@@ -0,0 +1,285 @@
+/*******************************************************************************
+ * Copyright 2017-2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/graph/node_builder.h"
+
+#include "ngraph_api.h"
+#include "ngraph_capture_variables.h"
+#include "ngraph_utils.h"
+
+using namespace std;
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+//
+// Utility function to check if placement on the NGRAPH device has been
+// requested.
+//
+// FIXME(amprocte): stubbed out for now because NGRAPH device is gone.
+//
+static bool NGraphPlacementRequested(const Node* node) { return true; }
+
+//
+// Main entry point for the variable-capture.
+//
+Status CaptureVariables(Graph* graph, std::vector<string> skip_these_nodes) {
+  if (config::IsEnabled() == false) {
+    return Status::OK();
+  }
+
+  std::vector<Node*> replaced_nodes;
+
+  for (auto node : graph->op_nodes()) {
+    if (NGraphPlacementRequested(node)) {
+      // Capture VariableV2
+      if (node->type_string() == "VariableV2") {
+        NGRAPH_VLOG(1) << "Capturing: " << node->name();
+
+        TensorShape shape;
+        DataType dtype;
+        TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "shape", &shape));
+        TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "dtype", &dtype));
+
+        std::string container;
+        std::string shared_name;
+        if (GetNodeAttr(node->attrs(), "container", &container) !=
+            Status::OK()) {
+          container = "";
+        }
+        if (GetNodeAttr(node->attrs(), "shared_name", &shared_name) !=
+            Status::OK()) {
+          shared_name = "";
+        }
+
+        Node* replacement;
+
+        // TODO(amprocte): Do we need to copy "_" attributes?
+        TF_RETURN_IF_ERROR(NodeBuilder(node->name(), "NGraphVariable")
+                               .Attr("shape", shape)
+                               .Attr("dtype", dtype)
+                               .Attr("container", container)
+                               .Attr("shared_name", shared_name)
+                               .Attr("ngraph_graph_id", 0)
```

Comment:
It is used as the variable op persists over the many graphs we get however the op behavior changes in each graph. So we need to attach a graph-id to differentiate between different occurrences of the same node (node name remains the same). It is hard coded to 0, it will be reset in rewrite_for_tracking.cc